tests/test_requests_compat.py

```diff
@@ -0,0 +1,2049 @@
+            (InvalidSchema, "localhost.localdomain:3128/"),
+            (InvalidSchema, "10.122.1.1:3128/"),
+            (InvalidURL, "http://"),
+        ),
+    )
+    def test_invalid_url(self, exception, url):
+        with pytest.raises(exception):
+            requests.get(url)
+
+    @pytest.mark.skip("httpx prepares requests by default")
+    def test_basic_building(self):
+        req = requests.Request()
+        req.url = "http://kennethreitz.org/"
+        req.data = {"life": "42"}
+
+        pr = req.prepare()
+        assert pr.url == req.url
+        assert pr.body == "life=42"
+
+    @pytest.mark.parametrize("method", ("GET", "HEAD"))
+    def test_no_content_length(self, httpbin, method):
+        req = requests.Request(method, httpbin(method.lower()))  # .prepare()
+        assert "Content-Length" not in req.headers
+
+    @pytest.mark.parametrize("method", ("POST", "PUT", "PATCH", "OPTIONS"))
+    def test_no_body_content_length(self, httpbin, method):
+        req = requests.Request(method, httpbin(method.lower()))
+        assert "content-length" not in req.headers
+
+    @pytest.mark.parametrize("method", ("POST", "PUT", "PATCH", "OPTIONS"))
+    def test_empty_content_length(self, httpbin, method):
+        req = requests.Request(method, httpbin(method.lower()), data="")
+        assert "content-length" not in req.headers
+
+    def test_override_content_length(self, httpbin):
+        headers = {"Content-Length": "not zero"}
+        r = requests.Request("POST", httpbin("post"), headers=headers)
+        assert "Content-Length" in r.headers
+        assert r.headers["Content-Length"] == "not zero"
+
+    def test_path_is_not_double_encoded(self):
+        request = requests.Request("GET", "http://0.0.0.0/get/test case")
+        assert request.url.path == "/get/test%20case"
+
+    @pytest.mark.parametrize(
+        "url, expected",
+        (
+            (
+                "http://example.com/path#fragment",
+                "http://example.com/path?a=b#fragment",
+            ),
+            pytest.param(
+                "http://example.com/path?key=value#fragment",
+                "http://example.com/path?key=value&a=b#fragment",
+                marks=pytest.mark.xfail(reason="HTTPX overrides existing query params"),
+            ),
+        ),
+    )
+    def test_params_are_added_before_fragment(self, url, expected):
+        request = requests.Request("GET", url, params={"a": "b"})
+        assert request.url == expected
+
+    def test_params_original_order_is_preserved_by_default(self):
+        param_ordered_dict = collections.OrderedDict(
+            (("z", 1), ("a", 1), ("k", 1), ("d", 1))
+        )
+        request = requests.Request(
+            "GET", "http://example.com/", params=param_ordered_dict
+        )
+        assert request.url == "http://example.com/?z=1&a=1&k=1&d=1"
+
+    @pytest.mark.xfail(reason="HTTPX doesn't support bytes-typed 'params' yet")
```

Comment:
Interesting! I didn't even know this was a supported case for Requests :) Should we support it too?